### ISSUE TITLE
CUDA OOM when using SFTTrainer with Phi-1.5B

### ISSUE THREAD

**Issue by SupreethRao99** [huggingface/trl#1034]
Hi,

I'm trying to supervised fine-tune a phi-1.5B model on a custom dataset with the SFTTrainer, my script closely follows the [sft_llama2.py](https://github.com/huggingface/trl/blob/main/examples/research_projects/stack_llama_2/scripts/sft_llama2.py). 

I'm training the model on 4x2080Ti (11G), the model 1.3B params should comfortable fit it the combined VRAM of the GPU's but I see a CUDA OOM errors when I start my training. 

my hyper parameters are as follows:
```python
    per_device_train_batch_size: Optional[int] = field(default=1, metadata={"help": "The batch size per GPU."})
    per_device_eval_batch_size: Optional[int] = field(default=1, metadata={"help": "The batch size per GPU for evaluation."})
    gradient_accumulation_steps: Optional[int] = field(default=8, metadata={"help": "The number of gradient accumulation steps."})
    gradient_checkpointing: Optional[bool] = field(default=False, metadata={"help": "Whether to use gradient checkpointing."})
```
furthermore , this is how I'm instantiating my model:
```python
self.base_model = AutoModelForCausalLM.from_pretrained(
            pretrained_model_name_or_path=self.script_args.model_name,
            quantization_config=self.bnb_config,
            device_map="auto",#{"": Accelerator().local_process_index},
            trust_remote_code=True,
            # torch_dtype=torch.float16,
            # use_flash_attention_2=False
        )
```

I cannot use PEFT or Gradient Checkpointing as Phi models are not supported.

**Comment by SupreethRao99** [comment]
using `Accelerator().local_process_index` will give OOM error while "auto" actually starts the training but takes up 35GB of VRAM with all 4 GPU's at about 35% utilisation

**Comment by tcapelle** [comment]
I managed to run Phi on a T4, but Phi currently does not support gradien cehckpoitning, so batch_size = 1 and max_seq_len of 512

**Comment by younesbelkada** [comment]
If you use phi models from @susnato https://huggingface.co/susnato/phi-1_5_dev you should be able to use GC without any problem

**Comment by susnato** [comment]
Yes, now that `Phi` is integrated into transformers, first please install transformers from the main, then you can run - 

```python 
from transformers import PhiForCausalLM

model = PhiForCausalLM.from_pretrained("susnato/phi-1_5_dev",  quantization_config=self.bnb_config) # the weights are currently being transferred to microsoft/phi-1_5
print(model.model.gradient_checkpointing) # False

# Enable gradient Checkpointing
model.gradient_checkpointing_enable()

print(model.model.gradient_checkpointing) # this should be True now.
 ```

If you want to freeze some internal layers alongside then please add this kwarg -

```python
model.gradient_checkpointing_enable(gradient_checkpointing_kwargs={"use_reentrant": False})
```

Even after enabling GC, if you don't see any reduction in the GPU memory then please tag me I will look into it.

**Comment by susnato** [comment]
The only downside of using `Phi` from the library is that you won't get support for `Flash Attention`. But it will be solved by this [PR](https://github.com/huggingface/transformers/pull/27661).
